Fixed #35546 -- Updated  contributing pages to align with current requirement of having an accepted ticket (#18341)
django/django: docs/internals/contributing/writing-code/submitting-patches.txt

Submitting contributions
========================

> We are always grateful for patches to Django's code.

Reviewer: A preference for the word "contributions" over "patches" was just merged in https://github.com/django/django/pull/18264, so please pull the latest and make sure this is not reverted. Thanks

To ensure a smooth process, please create a ticket on our `ticket tracker`_. before submitting your patch.

> Bug reports with associated tickets and contributions will be fixed *far* more quickly than those without.

Reviewer: I'd avoid suggesting that bugs are occasionally fixed without tickets. Typos and minor optimizations are usually merged without one, but we don't necessarily need to mention that -- I'd keep it clear that all behavior changes, including bugfixes, need accepted tickets.

Typo fixes and trivial documentation changes
============================================

If you are fixing a trivial issue, for example changing a word in the
documentation, the preferred process is to create a Trac ticket first.

> Once the ticket is accepted, you can then provide the patch using a GitHub pull request.

Reviewer: I think this part should largely stay as it was: I don't think we want to encourage tickets for typo fixes. Maybe we just need to clarify what kind of wording changes we mean. I just adjusted the wording of an error message in #18338, and I had a ticket because it was user-facing and changed the sense of the message. But if my change had only been to fix a typo, a ticket would have been unnecessary. I took the point of this section to be about _typo-like_ changes to docs that are uncontroversial in any possible way.

Submitting contributions
========================

We are always grateful for contributions to Django's code.
To ensure a smooth process, please create a ticket for :ref:`non-trivial-contribution`

> on our `ticket tracker`_ before submitting your contribution.

Reviewer 2: I'm tempted for us to not mention tickets yet and have something like
```
Thank you for your interest in contributing to Django's code.
```
(can be bigger mentioning that Django is run on the effort of volunteers and these docs are to help you understand Django's processes and land your contribution successfully)

Submitting contributions
========================

Thank you for your interest in contributing to Django's code.
Django is run on the effort of volunteers and these documentations
are to help you understand Django's processes and land your contribution

> successfully.

Reviewer 2: I think we haven't got the wording of this quite right (I appreciate I suggested this wording). 

I am now thinking, this is still true. I don't think this contradicts the advice of needing a ticket. This is a bug report with a solution (ie. someone raised a ticket in trac **and** raised a pull request). Although this is not yet accepted, when someone is trying to fix a bug, it is great to see the suggested fix (especially for regressions). This happens already quite well. 

So perhaps let's leave this paragraph as it was initially. @nessita what do you think?

Author: @sarahboyce should be this? 

> We're always grateful for contributions to Django's code. Indeed, bug reports
> with associated contributions will get fixed *far* more quickly than those
> without a solution.

Reviewer 2: Yes please

.. note::
    The Django software foundation requests that anyone contributing more than

>     a :ref:`trivial-change`, to Django sign and submit a `Contributor License

Reviewer 2: We can update the link text by doing...
```suggestion
    a :ref:`trivial change <trivial-change>`, to Django sign and submit a `Contributor License
```
(the line length will need to be adjusted)

  feature, the change should also contain documentation.

When you think your work is ready to be reviewed, send :doc:`a GitHub pull

> request <working-with-git>`.

Author: This was automatically updated from my VSCode. I am happy to revert it back if the change is unnecessary.

If you provide a non-trivial change, include evidence that alternatives have
been discussed on the `Django Forum`_ or |django-developers| list.
Additionally, ensure that a ticket has been submitted and accepted before

> proceeding with your patch.

Reviewer: The vicinity uses "contribution" rather than "patch".

If you provide a non-trivial change, include evidence that alternatives have
been discussed on the `Django Forum`_ or |django-developers| list.
Additionally, ensure that a ticket has been submitted and accepted before

> proceeding with your contribution.

Reviewer 2: I'm tempted to say, let's drop this addition here. 

I have two reasons to think this.

Firstly, just below here it says "... ask on the ticket for opinions" which implies there is a ticket

Secondly, I am very tempted to drop this whole "Non-trivial contributions" section all together. 
This was introduced in 407f62fd31689b6158902d3ffdcde64aa27679a7 (13 years ago) and hasn't really changed since. To me, this second classification of "trivial" vs "non-trivial" within the same page makes the docs harder to follow. I also think implying that "small bug fixes" are "trivial" is a little harsh on individuals who found it quite difficult to resolve these. 

So maybe let's not touch this section for now 🤔 
@nessita do you have opinions here?

Contribution checklist
======================

> Use this checklist to review to review a pull request.
> If you are reviewing a :ref:`non-trivial-contribution`,
> first ensure it has an accepted ticket.
> If the pull request passes all the criteria below and is not your own,
> please set the "Triage Stage" on the corresponding Trac ticket to "Ready for checkin".

Reviewer 2: ```suggestion
Use this checklist to review to review a pull request. If you are reviewing a
contribution that would not be :ref:`considered trivial <trivial-change>`,
first ensure it has an accepted ticket. If the pull request passes all the
criteria below and is not your own, please set the "Triage Stage" on the
corresponding Trac ticket to "Ready for checkin".
```

> Use this checklist to review to review a pull request.
> If you are reviewing a contribution that would not be
> :ref:`considered trivial <trivial-change>`, first ensure it has an accepted ticket.
> If the pull request passes all the criteria below and is not your own,
> please set the "Triage Stage" on the corresponding Trac ticket to "Ready for checkin".

Reviewer 2: If you want a new paragraph, it needs an empty line
We also need to wrap the docs at 80 characters 👍 
```suggestion
Use this checklist to review a pull request. If this contribution would not be
:ref:`considered trivial <trivial-change>`, first ensure it has an accepted
ticket before proceeding with the review.

If the pull request passes all the criteria below and is not your own, please
set the "Triage Stage" on the corresponding Trac ticket to "Ready for checkin".
```

----

**Note:** we could also arguably avoid this change as it says:
> Use this checklist to review a pull request. If you are reviewing a pull
request that is not your own and it passes all the criteria below

and we have added the requirement of the accepted ticket to the checklist.

However, I think in this case it's worth it because we are trying to suggest a volunteer doesn't invest too much effort in reviewing something Django may have no intention of accepting

If you've left comments for improvement on the pull request, please tick the
appropriate flags on the Trac ticket based on the results of your review:
"Patch needs improvement", "Needs documentation", and/or "Needs tests". As time
and interest permits, mergers do final reviews of "Ready for checkin" tickets
and will either commit the changes or bump it back to "Accepted" if further
work needs to be done.

If you're looking to become a member of the Review and Triage team, doing thorough reviews

> of contributions is a great way to earn trust.

Reviewer 2: Thank you for doing this, I think let's pull this out into a separate commit (or separate PR) as it more of a "side-quest" than what we're trying to achieve here 👍

Reviewer 2: Thank you, this also needs wrapping at 80 characters

"Patch needs improvement", "Needs documentation", and/or "Needs tests". As time
and interest permits, mergers do final reviews of "Ready for checkin" tickets
and will either commit the changes or bump it back to "Accepted" if further
work needs to be done.

.. note::
    Ensure that a ticket has been submitted and accepted before proceeding

>     with the review.

Reviewer: Thanks, but I can also imagine leaving this section as it was. Let's see what others think.

Reviewer 2: I think I agree as the second sentence onwards has the assumption that there is a ticket.

If we make a change here, I would update:
```
Use this checklist to review a pull request. If you are reviewing a pull
request that is not your own and it passes all the criteria below, please set
the "Triage Stage" on the corresponding Trac ticket to "Ready for checkin".
```
to something like
```
Use this checklist to review a pull request. If you are reviewing a non-trivial
pull request, first ensure the it has an accepted ticket. If the pull passes all 
the criteria below and is not your own, please set the "Triage Stage" on the 
corresponding Trac ticket to "Ready for checkin".
```
----

I noticed that these docs have the following sections:
1. "Typo fixes and trivial documentation changes": to me, this covers the contributions where we **do not expect a ticket**, pretty much all others we would expect an accepted ticket
2. "Non-trivial contributions": to me, this would normally mean contributions where we **expect a discussion** which concludes in an accepted ticket. However, in this context this seems to be saying that although we have accepted the ticket we still expect a discussion (whether that's the same or a new one) around the approach. We are not strict on this as a requirement, as alternative approaches are usually discussed in the review itself. This also doesn't mention DEPs and I'm not sure if that is what it is going for?

Unfortunately, that we have these two sections on the same page means that it's not very clear what we are talking about when mentioning a "non-trivial" thing.

I don't have the answer here but I would like us to link to the "Typo fixes and trivial documentation changes" section and when talking about a "non-trivial pull request" 👍 

----

Quite unrelated
> If you're looking to become a merger, doing thorough reviews of contributions is a great way to earn trust.

Perhaps "merger" might want to be updated to something like "member of the Review and Triage team" or something very generic like "hold a position of responsibility in Django". 
I appreciate that this is a separate concern to what this PR is addressing though 👍

If you're looking to become a merger, doing thorough reviews

> of patches is a great way to earn trust.

Reviewer: patches -> contributions

  :ref:`commit message format <committing-guidelines>`?
* Are you the patch author and a new contributor? Please add yourself to the
  :source:`AUTHORS` file and submit a `Contributor License Agreement`_.

> * If the change is a :ref:`non-trivial-contribution`, does the pull request have an accepted Trac ticket?

Reviewer 2: ```suggestion
* Does this have an accepted ticket on Trac? All contributions require a ticket
  unless the :ref:`change is considered trivial <trivial-change>`.
```
